h3-widgets/index.ts

```diff
@@ -0,0 +1,213 @@
+  const formula = await ws.getFormula({
+    column: selectedVariable,
+    operation: 'sum',
+    spatialFilter: getSpatialFilterFromViewState(viewState),
+    viewState,
+  })
+  formulaWidget.textContent = Intl.NumberFormat('en-US', {
+    maximumFractionDigits: 0,
+    // notation: 'compact'
+  }).format(formula.value)
+}
+
+const HISTOGRAM_WIDGET_ID = 'urbanity_widget';
+
+async function renderHistogram(ws: WidgetSource) {
+  histogramWidget.parentElement?.querySelector('.loader')?.classList.toggle('hidden', false);
+  histogramWidget.classList.toggle('hidden', true);
+
+  const categories = await ws.getCategories({
+    column: 'urbanity',
+    operation: 'count',
+    filterOwner: HISTOGRAM_WIDGET_ID,
+    spatialFilter: getSpatialFilterFromViewState(viewState),
+    viewState,
+  })
+
+  histogramWidget.parentElement?.querySelector('.loader')?.classList.toggle('hidden', true);
+  histogramWidget.classList.toggle('hidden', false);
+
+  const selectedCategory = filters['urbanity']?.[FilterType.IN]?.values[0];
+  const colors = categories.map((c) => c.name === selectedCategory ? 'rgba(255, 99, 132, 0.2)' : 'rgba(54, 162, 235, 0.2)')
+
+  if (histogramChart) {
+    histogramChart.data.labels = categories.map((c) => c.name);
+    histogramChart.data.datasets[0].data = categories.map((c) => c.value);
+    histogramChart.data.datasets[0].backgroundColor = colors;
+    histogramChart.update();
+  } else {
+    histogramChart = new Chart(histogramWidget, {
+      type: 'bar',
+      data: {
+        labels: categories.map((c) => c.name),
+        datasets: [
+          {
+            label: 'Urbanity category',
+            data: categories.map((c) => c.value),
+            backgroundColor: colors,
+          }
+        ]
+      },
+      options: {
+        onClick: async (ev, elems) => {
+          console.log(elems[0])
+          const index = elems[0]?.index
+          const categoryName = categories[index]?.name
+          if (!categoryName || categoryName === selectedCategory) {
+            removeFilter(filters, { column: 'urbanity' })
+          } else {
+            addFilter(filters, {
+              column: 'urbanity',
+              type: FilterType.IN,
+              values: [categoryName],
+              owner: HISTOGRAM_WIDGET_ID,
+            })
+          }
+          await render()
+        }
+      },
+    })
+  }
+}
+
+const debouncedRenderWidgets = debounce(renderWidgets, 500);
+
+// Main execution
+const map = new maplibregl.Map({
+  container: 'map',
+  style: BASEMAP.DARK_MATTER,
+  interactive: false
+});
+
+const deck = new Deck({
+  canvas: 'deck-canvas',
+  initialViewState: viewState,
+  controller: true
+});
+deck.setProps({
+  onViewStateChange: (props) => {
+    const {longitude, latitude, ...rest} = props.viewState;
+    map.jumpTo({center: [longitude, latitude], ...rest});
+    viewState = props.viewState;
+    debouncedRenderWidgets()
+  }
+});
+
```

Comment:
After zooming in to the southern tip part of Melilla, such that only one h3 bin is visible, I noticed that the total population in the tooltip doesn't match the formula widget:

<img width="1237" alt="Screenshot 2024-12-11 at 6 07 48 PM" src="https://github.com/user-attachments/assets/d7ff53f3-5583-4982-8527-253914077218" />

This might happen if the resolution of the tiles doesn't match what the Widgets API computes? Sorry this one is tricky to reproduce... just want to be sure we have the request parameters correct.

Reply:
This was also caused by the bug above. I think it might be resolved now.

Reply:
This still occurs for me, but I can see now that it's zoom-level dependent. If I zoom in enough, both the widget and the tile will agree on the "3666" result. If I start zooming out, the widget result increases (as if it's fetching a larger H3 cell?) first, and the size of the H3 cell displayed on the map does not change until I've zoomed much farther out.

Reply:
Hi Don! I'm tryong to replicate this and cannot make it. Maybe it is fetching some h3 cells that are behind the white panel ?

Reply:
If you have sometime soon we could do a quick meet with some live debugging

Reply:
I think it might be strictly a problem of the 'spatialFiltersResolution' changing at a time when the rendered tile H3 resolution has not changed. Here are two screenshots at slightly different zoom states, one matching and the other not:

### Matching
<img width="2709" alt="aligned" src="https://github.com/user-attachments/assets/deb0d82b-1226-4fbb-98be-e8d510eccf73" />

### Non-matching
<img width="2709" alt="misaligned" src="https://github.com/user-attachments/assets/1670732e-6318-460d-9d44-868df594b39a" />


Maybe a next step would be just to see if the same thing currently happens in C4R? If so then at least we know it isn't a new problem.

Reply:
After re-installing npm dependencies and getting latest `@carto/api-client`, unfortunately still seeing the same thing here.

Reply:
As discussed earlier today – the same mismatch happens in Builder/C4R as well:

[h3_mismatch.webm](https://github.com/user-attachments/assets/9f181900-25d6-41a0-bdb7-982746f631d9)

So whatever else is going on here, you've implemented the logic consistently with C4R! Assuming the Builder/C4R behavior is fine for now, I think we should just add a comment to `getSpatialFiltersResolution` in the `@carto/api-client` PR to keep track of this info, but it doesn't need to block the  changes for this initiative.

/cc @srtena @padawannn just FYI.

Reply:
For future reference the demo is using `carto-demo-data.demo_tables.derived_spatialfeatures_esp_h3res8_v1_yearly_v2`, just south of Melilla.

Reply:
Agreed to keep that as a known issue